python/tests/expression/test_literals.py

```diff
@@ -0,0 +1,665 @@
+    below_min_lit = small_lit.to(FloatType())
+    assert below_min_lit == BelowMin()
+
+
+@pytest.mark.parametrize(
+    "decimalType, decimalValue", [(DecimalType(9, 1), "34.6"), (DecimalType(9, 2), "34.56"), (DecimalType(9, 4), "34.5600")]
+)
+def test_double_to_decimal_conversion(decimalType, decimalValue):
+    lit = literal(34.56).to(DoubleType())
+
+    assert lit.to(decimalType).value.as_tuple() == Decimal(decimalValue).as_tuple()
+
+
+def test_decimal_to_decimal_conversion():
+    lit = literal(Decimal("34.11").quantize(Decimal(".01")))
+
+    assert lit.value.as_tuple() == lit.to(DecimalType(9, 2)).value.as_tuple()
+    assert lit.value.as_tuple() == lit.to(DecimalType(11, 2)).value.as_tuple()
+    assert lit.to(DecimalType(9, 0)) is None
+    assert lit.to(DecimalType(9, 1)) is None
+    assert lit.to(DecimalType(9, 3)) is None
+
+
+def test_timestamp_to_date():
+    epoch_lit = TimestampLiteral(EPOCH.timestamp() * 1000_000)
+    date_lit = epoch_lit.to(DateType())
+
+    assert date_lit.value == 0
+
+
+# STRING
+
+
+def test_string_literal():
+    sqrt2 = literal("1.414").to(StringType())
+    phi = literal("3.141").to(StringType())
+    phi_string_lit = StringLiteral("3.141")
+    phi_double_lit = literal(3.141).to(DoubleType())
+
+    assert sqrt2 != phi
+    assert phi != phi_double_lit
+    assert phi == phi_string_lit
+    assert phi == phi
+    assert sqrt2 < phi
+    assert sqrt2 <= phi
+    assert phi > sqrt2
+    assert phi >= sqrt2
+    assert str(phi) == '"3.141"'
+
+
+def test_string_to_string_literal():
+    assert literal("abc") == literal("abc").to(StringType())
+
+
+def test_string_to_date_literal():
+    date_str = literal("2017-08-18")
+    date = date_str.to(DateType())
+
+    avro_val = LOGICAL_WRITERS["int-date"](datetime.strptime("2017-08-18", "%Y-%m-%d"), None)
+    assert avro_val == date.value
+
+
+def test_string_to_time_literal():
+    time_str = literal("14:21:01.919")
+    time_lit = time_str.to(TimeType())
+
+    avro_val = LOGICAL_WRITERS["long-time-micros"](datetime.strptime("14:21:01.919", "%H:%M:%S.%f").time(), None)
+
+    assert avro_val == time_lit.value
+
+
+def test_string_to_timestamp_literal():
+    timestamp_str = literal("2017-08-18T14:21:01.919+00:00")
+    timestamp = timestamp_str.to(TimestamptzType())
+
+    avro_val = LOGICAL_WRITERS["long-timestamp-micros"](dateutil.parser.parse("2017-08-18T14:21:01.919+00:00"), None)
+    assert avro_val == timestamp.value
+
+    timestamp_str = literal("2017-08-18T14:21:01.919")
+    timestamp = timestamp_str.to(TimestampType())
+    assert avro_val == timestamp.value
+
+    timestamp_str = literal("2017-08-18T14:21:01.919-07:00")
+    timestamp = timestamp_str.to(TimestamptzType())
+    avro_val = LOGICAL_WRITERS["long-timestamp-micros"](dateutil.parser.parse("2017-08-18T21:21:01.919+00:00"), None)
+    assert avro_val == timestamp.value
+
+
+def test_timestamp_with_zone_without_zone_in_literal():
+    with raises(RuntimeError):
```

Comment:
Should this be `RuntimeError`? I think a more specific exception would be good so that it can be caught.